### PR TITLE
fix: don't create notification when already exists

### PR DESCRIPTION
**Pull request by MatousJobanek**
when the update of the status (that adds the condition) fails, then the controller creates two (or even more) "provisioned" notifications for the same user. I faced this problem when running e2e tests locally.

So I added a call that lists the notification by their labels - if there is one already present, then it sets only the condition, but doesn't create the notification CR.

### CODE REVIEW

File: controllers/masteruserrecord/sync.go [hunk 1]
```diff
@@ -183,32 +183,41 @@ func (s *Synchronizer) alignReadiness() (bool, error) {
...
+			toolchainv1alpha1.NotificationUserNameLabelKey: s.record.Name,
+			// NotificationTypeLabelKey is only used for easy lookup for debugging and e2e tests
+			toolchainv1alpha1.NotificationTypeLabelKey: toolchainv1alpha1.NotificationTypeProvisioned,
+		}
+		opts := client.MatchingLabels(labels)
+		notificationList := &toolchainv1alpha1.NotificationList{}
+		if err := s.hostClient.List(context.TODO(), notificationList, opts); err != nil {
 			return false, err
 		}
+		// if there is no existing notification with that labels
```
Comment (alexeykazakov) [inline controllers/masteruserrecord/sync.go:197]:
```suggestion
		// if there is no existing notification with these labels
```
```suggestion
		// if there is no existing notification with that labels
```

File: controllers/masteruserrecord/sync.go [hunk 2]
```diff
@@ -183,32 +183,41 @@ func (s *Synchronizer) alignReadiness() (bool, error) {
...
-				// (or removed) in a future PR
-				UserID:   s.record.Labels[toolchainv1alpha1.MasterUserRecordOwnerLabelKey],
-				Template: notificationtemplates.UserProvisioned.Name,
-			},
-		}
-
-		err := controllerutil.SetControllerReference(s.record, notification, s.scheme)
-		if err != nil {
+		labels := map[string]string{
+			// NotificationUserNameLabelKey is only used for easy lookup for debugging and e2e tests
```
Comment (alexeykazakov) [inline controllers/masteruserrecord/sync.go:187]: It is not for tests or debugging anymore. You do rely on this label in the controller now.

File: controllers/masteruserrecord/sync.go [hunk 3]
```diff
@@ -183,32 +183,41 @@ func (s *Synchronizer) alignReadiness() (bool, error) {
...
+			// NotificationTypeLabelKey is only used for easy lookup for debugging and e2e tests
+			toolchainv1alpha1.NotificationTypeLabelKey: toolchainv1alpha1.NotificationTypeProvisioned,
+		}
+		opts := client.MatchingLabels(labels)
+		notificationList := &toolchainv1alpha1.NotificationList{}
+		if err := s.hostClient.List(context.TODO(), notificationList, opts); err != nil {
 			return false, err
 		}
+		// if there is no existing notification with that labels
+		if len(notificationList.Items) == 0 {
```
Comment (alexeykazakov) [inline controllers/masteruserrecord/sync.go:198]: Let's log an info message saying that the notification was not created because it's already exists.

**Review by alexeykazakov — APPROVED** [review]
Looks good. Just a few minor comments.

File: controllers/masteruserrecord/sync.go (hunk 2 above)
Comment (MatousJobanek) [reply controllers/masteruserrecord/sync.go]: right :+1: 
I just blindly copy-pasted it and didn't pay attention on the comments - fixed in ea4f95c

File: controllers/masteruserrecord/sync.go (hunk 3 above)
Comment (MatousJobanek) [reply controllers/masteruserrecord/sync.go:196]: done 425e620

**Pull request merged**